core/drivers/crypto/se050/adaptors/utils/scp_config.c

```diff
@@ -402,34 +294,88 @@ static sss_status_t get_config_key(struct se050_scp_key *keys __maybe_unused)
 #endif
 }
 
-sss_status_t se050_scp03_get_keys(struct se050_scp_key *keys)
+sss_status_t se050_scp03_subkey_derive(struct se050_scp_key *keys)
 {
-	sss_status_t (*get_keys[])(struct se050_scp_key *) = {
-		&get_config_key, &get_db_key, &get_ofid_key,
+	struct {
+		const char *name;
+		uint8_t *data;
+	} key[3] = {
+		[0] = { .name = "dek", .data = keys->dek },
+		[1] = { .name = "mac", .data = keys->mac },
+		[2] = { .name = "enc", .data = keys->enc },
 	};
+	uint8_t msg[SE050_SCP03_KEY_SZ + 3] = { 0 };
 	size_t i = 0;
 
-	for (i = 0; i < ARRAY_SIZE(get_keys); i++)
-		if ((*get_keys[i])(keys) == kStatus_SSS_Success)
-			return kStatus_SSS_Success;
+	if (tee_otp_get_die_id(msg + 3, SE050_SCP03_KEY_SZ))
+		return kStatus_SSS_Fail;
 
-	return kStatus_SSS_Fail;
+	for (i = 0; i < ARRAY_SIZE(key); i++) {
+		memcpy(msg, key[i].name, 3);
+		if (huk_subkey_derive(HUK_SUBKEY_SE050, msg, sizeof(msg),
+				      key[i].data, SE050_SCP03_KEY_SZ))
+			return kStatus_SSS_Fail;
+	}
+
+	return kStatus_SSS_Success;
 }
 
-sss_status_t se050_scp03_put_keys(struct se050_scp_key *keys,
-				  struct se050_scp_key *cur_keys)
+bool se050_scp03_enabled(void)
+{
+	return scp03_enabled;
+}
 
+void se050_scp03_set_enable(enum se050_scp03_ksrc ksrc)
 {
-	sss_status_t status = kStatus_SSS_Success;
+	scp03_enabled = true;
+	scp03_ksrc = ksrc;
+}
 
-	if (cur_keys) {
-		status = se050_scp03_get_keys(cur_keys);
-		if (status != kStatus_SSS_Success)
-			return status;
-	}
+void se050_scp03_set_disable(void)
+{
+	scp03_enabled = false;
+}
 
-	if (scp03db_write_keys(keys))
+sss_status_t se050_scp03_get_keys(struct se050_scp_key *keys,
+				  enum se050_scp03_ksrc ksrc)
+{
+	switch (ksrc) {
+	case SCP03_CFG:
+		return get_config_key(keys);
+	case SCP03_DERIVED:
+		return se050_scp03_subkey_derive(keys);
+	case SCP03_OFID:
+		return get_ofid_key(keys);
+	default:
 		return kStatus_SSS_Fail;
+	}
+}
 
-	return kStatus_SSS_Success;
+sss_status_t se050_scp03_get_current_keys(struct se050_scp_key *keys)
+{
+	if (se050_scp03_enabled())
+		return se050_scp03_get_keys(keys, scp03_ksrc);
+
+	return kStatus_SSS_Fail;
 }
+
+void se050_scp03_show_keys(struct se050_scp_key *new_keys,
+			   struct se050_scp_key *cur_keys)
+{
+	if (!IS_ENABLED(CFG_CORE_SE05X_DISPLAY_SCP03_KEYS))
+		return;
+
+	if (!new_keys || !cur_keys)
```

Comment:
Why allowing this function to be called with NULL pointers? It seems it would be an error to do so.

Reply:
right...so you want me to remove these checks?

Reply:
I'll just remove the function entirely...dont need a function that is called from just one place